#####################################################################################
###### Script that generates the linear synteny plot (Figure 1 in Teterina et al. 2020)
###### Input files: coordinates of 1-to-1 ortologs, chromosome sizes, and protein identities

setwd("/Users/anastasia/Documents/Phillips_lab/drafts/CR_genome/Synteny")

library(ggplot2)

#files with coordinates of orthologs (in the same order)
CR<-read.csv("CR_SYNTENY.txt",header=F,sep="\t") #C. remanei orthologs

#chromosome start end
#II	10484428	10480858
#IV	70427	68683
#V	89564	91003
#V	98430	99972
#V	102370	107467
#V	111668	110537

CE<-read.csv("CE_SYNTENY.txt",header=F,sep="\t") # C.elegans orthologs

#chromosome start end, in the orthogroup order
#chrII_pilon	15511838	15519315
#chrIV_pilon	17751743	17747835
#chrV_pilon	10411930	10413242
#chrV_pilon	10427531	10429085
#chrV_pilon	10431776	10436878
#chrV_pilon	10438606	10436905

ID<-read.csv("CR_CE_identity_short.txt",header=F,sep=" ") #global identity of protein sequences generated by https://gist.github.com/avrilcoghlan/5311008

#C.remanei_gene_ID C.elegans_gene_ID %Identity
#CRPC05910.1 WBGene00007064 77.5
#CRPC12047.1 WBGene00003525 69.2
#CRPC16515.1 WBGene00005532 74.1
#CRPC16518.1 WBGene00004964 83.3
#CRPC16519.1 WBGene00007071 85.7
#CRPC16522.1 WBGene00000724 73.3


#Chromosome sizes, C. remanei
cr1=17247545
cr2=19935723
cr3=17877849
cr4=25790997
cr5=22502457
cr6=21501900

#Chromosome sizes, C. elegans
ce1=15331301
ce2=15525148
ce3=14108536
ce4=17759200
ce5=21243235
ce6=18110855


#Combine files

link11<-cbind(CE,CR)
link11<-cbind(link11,ID$V3)

colnames(link11)<-c("ce","start1","end1","cr","start2","end2","identity")
link11$ce<-gsub("_pilon","",link11$ce)
link11$ce<-gsub("chr","",link11$ce)
link11$cr<-gsub("^","CR-",link11$cr)

gap<-2500000 #distance between chromosomes on the Figure

LINEAR<-link11[link11$identity>50,]



#Change all coordinates
#C. remanei

LINEAR[LINEAR$cr=="CR-II",]$start2<-LINEAR[LINEAR$cr=="CR-II",]$start2 + gap + cr1
LINEAR[LINEAR$cr=="CR-II",]$end2<-LINEAR[LINEAR$cr=="CR-II",]$end2 + gap + cr1

LINEAR[LINEAR$cr=="CR-III",]$start2<-LINEAR[LINEAR$cr=="CR-III",]$start2 + 2*gap + cr1 + cr2
LINEAR[LINEAR$cr=="CR-III",]$end2<-LINEAR[LINEAR$cr=="CR-III",]$end2 + 2*gap + cr1 + cr2

LINEAR[LINEAR$cr=="CR-IV",]$start2<-LINEAR[LINEAR$cr=="CR-IV",]$start2 + 3*gap + cr1 + cr2 + cr3
LINEAR[LINEAR$cr=="CR-IV",]$end2<-LINEAR[LINEAR$cr=="CR-IV",]$end2 + 3*gap + cr1 + cr2 + cr3

LINEAR[LINEAR$cr=="CR-V",]$start2<-LINEAR[LINEAR$cr=="CR-V",]$start2 + 4*gap + cr1 + cr2 + cr3 + cr4
LINEAR[LINEAR$cr=="CR-V",]$end2<-LINEAR[LINEAR$cr=="CR-V",]$end2 + 4*gap + cr1 + cr2 + cr3 + cr4

LINEAR[LINEAR$cr=="CR-X",]$start2<-LINEAR[LINEAR$cr=="CR-X",]$start2 + 5*gap + cr1 + cr2 + cr3 + cr4 + cr5
LINEAR[LINEAR$cr=="CR-X",]$end2<-LINEAR[LINEAR$cr=="CR-X",]$end2 + 5*gap + cr1 + cr2 + cr3 + cr4 + cr5

#C. elegans

LINEAR[LINEAR$ce=="I",]$start1<-LINEAR[LINEAR$ce=="I",]$start1 + round((cr1-ce1)/2)
LINEAR[LINEAR$ce=="I",]$end1<-LINEAR[LINEAR$ce=="I",]$end1 + round((cr1-ce1)/2)

LINEAR[LINEAR$ce=="II",]$start1<-LINEAR[LINEAR$ce=="II",]$start1 + (cr1-ce1) + gap + round((cr2-ce2)/2) + ce1
LINEAR[LINEAR$ce=="II",]$end1<-LINEAR[LINEAR$ce=="II",]$end1 + (cr1-ce1) + gap + round((cr2-ce2)/2) + ce1

LINEAR[LINEAR$ce=="III",]$start1<-LINEAR[LINEAR$ce=="III",]$start1 + (cr1-ce1) + (cr2-ce2) + 2*gap + round((cr3-ce3)/2) + ce1 + ce2
LINEAR[LINEAR$ce=="III",]$end1<-LINEAR[LINEAR$ce=="III",]$end1 + (cr1-ce1) + (cr2-ce2) + 2*gap + round((cr3-ce3)/2) + ce1 + ce2

LINEAR[LINEAR$ce=="IV",]$start1<-LINEAR[LINEAR$ce=="IV",]$start1 + (cr1-ce1) + (cr2-ce2) + (cr3-ce3) + 3*gap + round((cr4-ce4)/2) + ce1 + ce2 + ce3
LINEAR[LINEAR$ce=="IV",]$end1<-LINEAR[LINEAR$ce=="IV",]$end1 + (cr1-ce1) + (cr2-ce2) + (cr3-ce3) + 3*gap + round((cr4-ce4)/2) + ce1 + ce2 + ce3

LINEAR[LINEAR$ce=="V",]$start1<-LINEAR[LINEAR$ce=="V",]$start1 + (cr1-ce1) + (cr2-ce2) + (cr3-ce3) + (cr4-ce4) + 4*gap + round((cr5-ce5)/2) + ce1 + ce2 + ce3 + ce4
LINEAR[LINEAR$ce=="V",]$end1<-LINEAR[LINEAR$ce=="V",]$end1 + (cr1-ce1) + (cr2-ce2) + (cr3-ce3) + (cr4-ce4) +4*gap + round((cr5-ce5)/2) + ce1 + ce2 + ce3 + ce4

LINEAR[LINEAR$ce=="X",]$start1<-LINEAR[LINEAR$ce=="X",]$start1 + (cr1-ce1) + (cr2-ce2) + (cr3-ce3)+ (cr4-ce4) + (cr5-ce5) + 5*gap + round((cr6-ce6)/2) + ce1 + ce2 + ce3 + ce4 + ce5
LINEAR[LINEAR$ce=="X",]$end1<-LINEAR[LINEAR$ce=="X",]$end1 + (cr1-ce1) + (cr2-ce2) + (cr3-ce3) + (cr4-ce4) + (cr5-ce5) + 5*gap + round((cr6-ce6)/2) + ce1 + ce2 + ce3 + ce4 + ce5



#Boundaries of chromosomes

start=c(1,gap + cr1,2*gap + cr1 + cr2,3*gap + cr1 + cr2 + cr3,4*gap + cr1 + cr2 + cr3 + cr4, 5*gap + cr1 + cr2 + cr3 + cr4 + cr5)
end=c(cr1,gap + cr1 + cr2,2*gap + cr1 + cr2 +cr3,3*gap + cr1 + cr2 + cr3 + cr4,4*gap + cr1 + cr2 + cr3 + cr4 + cr5, 5*gap + cr1 + cr2 + cr3 + cr4 + cr5 + cr6)
diff=c((cr1-ce1), (cr2-ce2), (cr3-ce3), (cr4-ce4), (cr5-ce5), cr6-ce6)

#change positions of chromosomes, sp parameter is the distance between C. remanei and C.elegans chromosomes on the plot
rect<-data.frame(start=c(start,start+diff/2),end=c(end, end-diff/2),sp=c(rep(1,6),rep(10000,6)))

#chromosome labels
chrlabel<-data.frame(pos=c((start+end)/2),sp=c(rep(1,6),rep(10000,6)),text=c("I","II","III","IV","V","X","I","II","III","IV","V","X"))


LINEAR$ID<-row.names(LINEAR)

#Orientation of genes
LINEAR$COLOR<-"SAME"
LINEAR[LINEAR$start1>LINEAR$end1 & LINEAR$start2<LINEAR$end2,]$COLOR<-"DIFF"
LINEAR[LINEAR$start1<LINEAR$end1 & LINEAR$start2>LINEAR$end2,]$COLOR<-"DIFF"

LINEAR_df2<-data.frame(COORD=c((LINEAR$start1 + LINEAR$end1)/2,(LINEAR$start2 + LINEAR$end2)/2), ID=c(LINEAR$ID,LINEAR$ID), CHROM=c(as.character(LINEAR$cr),as.character(LINEAR$cr)),SP=c(rep(10000,nrow(LINEAR)),rep(1,nrow(LINEAR))),COLOR=c(LINEAR$COLOR,LINEAR$COLOR), stringsAsFactors = FALSE)


######################################################################
######################## Figure 1 ####################################
######################################################################

ggplot(LINEAR_df2, aes(y = COORD)) + geom_line(aes(x = SP, group = ID,color=COLOR),alpha=0.3,size=0.1) +
      coord_flip() +
       theme_void() +
       theme(legend.position = "none") +
       geom_rect(data=rect, aes(ymin=start, ymax=end, xmin=rect$sp-300, xmax=rect$sp+300), inherit.aes = F,fill="grey") +
       geom_text(data=chrlabel, inherit.aes = FALSE, aes(y=pos, x=c(chrlabel$sp[1:6]-1000,chrlabel$sp[7:12]+1000),label=text)) +
       theme(plot.margin = margin(0.2, 0, 0.2, 0, "cm")) +
       geom_segment(data=LINEAR_df2[LINEAR_df2$SP==10000,], aes(y = COORD, yend=COORD, x=10000-300, xend=10000+300,color=COLOR), inherit.aes = F,alpha=0.1)  +
       geom_segment(data=LINEAR_df2[LINEAR_df2$SP==1,], aes(y = COORD, yend=COORD, x=1-300, xend=1+300,color=COLOR), inherit.aes = F,alpha=0.1) +
       scale_colour_manual(values = c("#FFBD58","#2F9DA1")) ->s

ggsave("Synteny_CR_CE.pdf",s, width=8, height=2.2, units="in", scale=0.8)
